File: learnLeetCode/Day20/Part1/text_editor.py
# LC 2296 — Design a Text Editor
# Two stacks: left (text left of cursor), right (text right of cursor)

class TextEditor:
    def __init__(self):
        self.left = []   # list of chars
        self.right = []  # list of chars (top is next char to the right)

    def addText(self, text: str) -> None:
        for ch in text:
            self.left.append(ch)

    def deleteText(self, k: int) -> int:
        cnt = 0
        while self.left and k > 0:
            self.left.pop()
            k -= 1
            cnt += 1
        return cnt

    def _last10(self) -> str:
        # return up to last 10 chars on the left side of cursor
        L = len(self.left)
        start = max(0, L - 10)
        return ''.join(self.left[start:L])

    def cursorLeft(self, k: int) -> str:
        while k > 0 and self.left:
            self.right.append(self.left.pop())
            k -= 1
        return self._last10()

    def cursorRight(self, k: int) -> str:
        while k > 0 and self.right:
            self.left.append(self.right.pop())
            k -= 1
        return self._last10()

if __name__ == "__main__":
    ed = TextEditor()
    ed.addText("leetcode")
    print(ed.deleteText(4))       # 4
    ed.addText("practice")
    print(ed.cursorRight(3))      # etpractice
    print(ed.cursorLeft(8))       # leet
    print(ed.deleteText(10))      # 4

File: learnLeetCode/Day20/Part1/text_editor.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 2296 — Design a Text Editor
// Two stacks: left of cursor, right of cursor

class TextEditor {
    vector<char> left, right; // right.back() is next char to the right
public:
    TextEditor() {}

    void addText(string text){
        for(char c: text) left.push_back(c);
    }

    int deleteText(int k){
        int cnt = 0;
        while(!left.empty() && k>0){ left.pop_back(); --k; ++cnt; }
        return cnt;
    }

    string last10(){
        int L = (int)left.size();
        int start = max(0, L-10);
        string out; out.reserve(min(10, L));
        for(int i=start;i<L;++i) out.push_back(left[i]);
        return out;
    }

    string cursorLeft(int k){
        while(k>0 && !left.empty()){
            right.push_back(left.back()); left.pop_back(); --k;
        }
        return last10();
    }

    string cursorRight(int k){
        while(k>0 && !right.empty()){
            left.push_back(right.back()); right.pop_back(); --k;
        }
        return last10();
    }
};

int main(){
    TextEditor ed;
    ed.addText("leetcode");
    cout << ed.deleteText(4) << "\n"; // 4
    ed.addText("practice");
    cout << ed.cursorRight(3) << "\n"; // etpractice
    cout << ed.cursorLeft(8) << "\n";  // leet
    cout << ed.deleteText(10) << "\n"; // 4
    return 0;
}
